actionpack/lib/action_dispatch/middleware/app_info.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module ActionDispatch
+  # Middleware that adds application version and environment information to response headers.
+  # This helps with debugging, deployment verification, and monitoring.
+  #
+  # The middleware can be configured through Rails.application.config.app_version:
+  #
+  #   config.app_version.enabled = true  # Enable/disable the middleware
+  #   config.app_version.add_headers = true  # Add version headers to responses
+  #   config.app_version.include_revision = -> { Rails.env.local? }  # Include git revision
+  #
+  # By default, it adds the following headers:
+  # * X-App-Version: The application version
+  # * X-App-Environment: The current environment
+  #
+  class AppInfo
+    def initialize(app, options = {})
+      @app = app
+      @version_header = options[:version_header] || "X-App-Version"
+      @environment_header = options[:environment_header] || "X-App-Environment"
+    end
+
+    def call(env)
+      status, headers, response = @app.call(env)
+
+      if Rails.application.config.app_version.add_headers
+        headers[@version_header] = version_string
+        headers[@environment_header] = Rails.application.app_environment.to_s
+      end
```

Comment:
Rather than a middleware that is a noop based on config, we should only insert the middleware if the config is enabled.